datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -283,6 +283,11 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
     Ok(match array {
         ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(v, tz_opt)) => {
             let nano = (f)(*v)?;
+
+            if nano.is_none() {
+                return Ok(ColumnarValue::Scalar(ScalarValue::Null));
```

Comment:
The issue is because you use `ScalarValue::Null` here.

You can take a look of `enum ScalarValue`. `ScalarValue::Null` is for `DataType::Null`. For other types, they can take a `Some(value)` or a `None` which represents a null value for the type.

For `TimestampNanosecond` type, you just need to put `ScalarValue::TimestampNanosecond` with `None` here.

Reply:
This is how `TimestampNanosecond` is defined in `enum ScalarValue`:

```
/// Timestamp Nanoseconds
TimestampNanosecond(Option<i64>, Option<Arc<str>>),
```

The first argument is `Option<i64>`. If it is a `None`, meaning that is a null `TimestampNanosecond` scalar value.